packages/aws-cdk/toolkit/lib/cli-io-host.ts

```diff
@@ -0,0 +1,143 @@
+import * as chalk from 'chalk';
+
+/**
+ * Basic message structure for toolkit notifications.
+ * Messages are emitted by the toolkit and handled by the IoHost.
+ */
+interface IoMessage {
+  /**
+   * @param time The time the message was emitted.
+   */
+  readonly time: Date;
+
+  /**
+   * @param level The log level of the message.
+   */
+  readonly level: IoMessageLevel;
+
+  /**
+   * @param action The action that triggered the message.
+   */
+  readonly action: IoAction;
+
+  /**
+   * @param code A short code to identify the message.
+   */
+  readonly code: string;
+
+  /**
+   * @param message The message text.
+   */
+  readonly message: string;
+
+  /**
+   * @param forceStdout If true, the message will be written to stdout regardless of log level.
+   *
+   * @default false
+   */
+  readonly forceStdout?: boolean;
+}
+
+export type IoMessageLevel = 'error' | 'warn' | 'info' | 'debug' | 'trace';
+
+export type IoAction = 'synth' | 'list' | 'deploy' | 'destroy';
+
+/**
+ * Options for the CLI IO host.
+ */
+interface CliIoHostOptions {
+  /**
+   * @param useTTY If true, the host will use TTY features like color.
+   */
+  useTTY?: boolean;
+
+  /**
+   * @param ci Flag representing whether the current process is running in a CI environment,
+   * If true, the host will write all messages to stdout.
+   *
+   * @default false
+   */
+  ci?: boolean;
+}
+
+/**
+ * A simple IO host for the CLI that writes messages to the console.
+ */
+export class CliIoHost {
+  private readonly pretty_messages: boolean;
+  private readonly ci: boolean;
+
+  constructor(options: CliIoHostOptions) {
+    this.pretty_messages = options.useTTY ?? process.stdout.isTTY ?? false;
+    this.ci = options.ci ?? false;
+  }
+
+  /**
+   * Notifies the host of a message.
+   * The caller waits until the notification completes.
+   */
+  async notify(msg: IoMessage): Promise<void> {
+    const output = this.formatMessage(msg);
+
+    const stream = this.getStream(msg.level, msg.forceStdout ?? false);
+
+    return new Promise((resolve, reject) => {
+      stream.write(output, (err) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  /**
+   * Determines which output stream to use based on log level and configuration.
+   */
+  private getStream(level: IoMessageLevel, forceStdout: boolean) {
+    /**
+     * For legacy purposes all log streams are written to stderr by default, unless
+     * specified otherwise, by passing `forceStdout`, which is used by the `data()` logging function, or
+     * if the CDK is running in a CI environment. This is because some CI environments will immediately
+     * fail if stderr is written to. In these cases, we detect if we are in a CI environment and
+     * write all messages to stdout instead.
+     */
```

Comment:
should be regular doc

Reply:
What do you mean by regular doc? Do you mean `//`?

Reply:
yes